LightGBM/data.py

```diff
@@ -0,0 +1,72 @@
+import numpy as np
+import pandas as pd
+
+
+def downcast(df, verbose=False):
+    """
+    Downcasts the data to reduce memory usage.
+    Adapted from: https://www.kaggle.com/kyakovlev/m5-simple-fe
+
+    Args:
+        df = [pd.DataFrame] pandas dataframe
+    
+    Returns [pd.DataFrame]:
+        Downcasted data.
+    """
+    numerics = ['int16', 'int32', 'int64', 'float16', 'float32', 'float64']
+    start_mem = df.memory_usage().sum() / 1024**2 
+
+    for col in df.columns:
+        col_type = df[col].dtypes
+        if col_type in numerics:
+            c_min = df[col].min()
+            c_max = df[col].max()
+            if str(col_type)[:3] == 'int':
+                if c_min > np.iinfo(np.int8).min and c_max < np.iinfo(np.int8).max:
+                    df[col] = df[col].astype(np.int8)
+                elif c_min > np.iinfo(np.int16).min and c_max < np.iinfo(np.int16).max:
+                       df[col] = df[col].astype(np.int16)
+                elif c_min > np.iinfo(np.int32).min and c_max < np.iinfo(np.int32).max:
+                    df[col] = df[col].astype(np.int32)
+                elif c_min > np.iinfo(np.int64).min and c_max < np.iinfo(np.int64).max:
+                    df[col] = df[col].astype(np.int64)  
+            else:
+                if c_min > np.finfo(np.float16).min and c_max < np.finfo(np.float16).max:
+                    df[col] = df[col].astype(np.float16)
+                elif c_min > np.finfo(np.float32).min and c_max < np.finfo(np.float32).max:
+                    df[col] = df[col].astype(np.float32)
+                else:
+                    df[col] = df[col].astype(np.float64)   
+        elif col == 'date':
+            df[col] = pd.to_datetime(df[col], format='%Y-%m-%d')
+        elif str(col_type) == 'object':
+            df[col] = df[col].astype('category')
+        else:
+            print(f'Unexpected dtype {col_type} in dataframe.')   
+
+    end_mem = df.memory_usage().sum() / 1024**2
+    if verbose: 
+        print('Mem. usage decreased to {:5.2f} Mb ({:.1f}% reduction)'.format(end_mem, 100 * (start_mem - end_mem) / start_mem))
```

Comment:
You should read up on f-strings. They are very powerful for situations like this one.

`print('Mem. usage decreased to {end_mem:5.2f} Mb ({(start_mem - end_mem) / start_mem:.1%} reduction)')`

Notice that you do not need to `100 *` to compute the percentage.

Reply:
I do use f-string elsewhere in the code. I copy-pasted this, but I probably should have changed it in the first place. I did not know you can show percentages like this with f-strings, so thanks for that 🙂.